Lib/test/test_functools.py

```diff
@@ -2905,6 +2905,88 @@ def _(arg: typing.List[float] | bytes):
         self.assertEqual(f(""), "default")
         self.assertEqual(f(b""), "default")
 
+    def test_type_argument(self):
+        @functools.singledispatch
+        def f(arg):
+            return "default"
+
+        @f.register
+        def _(arg: type[int]):
+            return "type[int]"
+
+        @f.register
+        def _(arg: typing.Type[float]):
+            return "type[float]"
+
+        @f.register(type[str])
+        def _(arg):
+            return "type[str]"
+
+        @f.register(typing.Type[bytes])
+        def _(arg):
+            return "type[bytes]"
+
+        self.assertEqual(f(int), "type[int]")
+        self.assertEqual(f(float), "type[float]")
+        self.assertEqual(f(str), "type[str]")
+        self.assertEqual(f(bytes), "type[bytes]")
+        self.assertEqual(f(2), "default")
+
+    def test_type_argument_mro(self):
+        class A:
+            pass
+
+        class B(A):
+            pass
+
+        class C:
+            pass
+
+        @functools.singledispatch
+        def f(arg):
+            return "default"
+
+        @f.register
+        def _(arg: type[A]):
+            return "type[A]"
+
+        @f.register
+        def _(arg: B):
+            return "B"
+
+        self.assertEqual(f(B), "type[A]")
+        self.assertEqual(f(C), "default")
+
+    def test_type_argument_unions(self):
+        @functools.singledispatch
+        def f(arg):
+            return "default"
+
+        @f.register
+        def _(arg: type[list|dict]):
+            return "type[list|dict]"
+
+        @f.register
+        def _(arg: type[set]|typing.Type[type(None)]):
+            return "type[set]|type[NoneType]"
+
+        self.assertEqual(f(list), "type[list|dict]")
+        self.assertEqual(f(type(None)), "type[set]|type[NoneType]")
```

Comment:
I'm wondering if treating `type[a|b]` and `type[a]|type[b]` the same actually makes sense and even if so, whether union types should be supported for `type[...]` arguments at all:

E.g. if we have a single-dispatch function with an implementation for `type[a|b|c]`, wouldn't people expect to be able to pass e.g. `a|b` to it and have it dispatch to that implementation? That would be much harder to implement than my current naive implementation of just splitting up unions into their individual constituent types, especially considering `issubclass(a|b, a|b|c)` isn't even possible.

So maybe unions of `type[...]`s and `type[...]`s of unions should just not be allowed for now, deferring them to when (if ever) `issubclass` supports these kinds of checks (or Python gets another `issubtype` function).

Reply:
OTOH, it's unlikely that introducing support for dispatching to `type[a|b|c]` given `a|b` later on would be a breaking change... If only dispatching on single types is supported for now then that is all people will use, nobody will rely on the fact that `a|b` dispatches to the default implementation, they'll just never pass `a|b` to a single-dispatch function I should think. So it might be fine to implement it like this for now and leave only the "proper" handling for later.